near-metrics/src/lib.rs

```diff
@@ -0,0 +1,143 @@
+//! Because we use a global registry there is no namespace per crate, it's one big global space.
+//!
+//! See the [Prometheus naming best practices](https://prometheus.io/docs/practices/naming/) when
+//! choosing metric names.
+//!
+//! ## Example
+//!
+//! ```rust
+//! #[macro_use]
+//! extern crate lazy_static;
+//! use near_metrics::*;
+//!
+//! // These metrics are "magically" linked to the global registry PEER_CONNECTIONS_TOTALdefined in `lighthouse_metrics`.
+//! lazy_static! {
+//!     pub static ref RUN_COUNT: Result<IntCounter> = try_create_int_counter(
+//!         "runs_total",
+//!         "Total number of runs"
+//!     );
+//!     pub static ref CURRENT_VALUE: Result<IntGauge> = try_create_int_gauge(
+//!         "current_value",
+//!         "The current value"
+//!     );
+//!     pub static ref RUN_TIME: Result<Histogram> =
+//!         try_create_histogram("run_seconds", "Time taken (measured to high precision)");
+//! }
+//!
+//!
+//! fn main() {
+//!     for i in 0..100 {
+//!         inc_counter(&RUN_COUNT);
+//!         let timer = start_timer(&RUN_TIME);
+//!
+//!         for j in 0..10 {
+//!             set_gauge(&CURRENT_VALUE, j);
+//!             println!("Howdy partner");
+//!         }
+//!
+//!         stop_timer(timer);
+//!     }
+//! }
+//! ```
+
+use prometheus::{HistogramOpts, HistogramTimer, Opts};
+
+pub use prometheus::{Encoder, Histogram, IntCounter, IntGauge, Result, TextEncoder};
+
+/// Collect all the metrics for reporting.
+pub fn gather() -> Vec<prometheus::proto::MetricFamily> {
+    prometheus::gather()
+}
+
+/// Attempts to crate an `IntCounter`, returning `Err` if the registry does not accept the counter
+/// (potentially due to naming conflict).
+pub fn try_create_int_counter(name: &str, help: &str) -> Result<IntCounter> {
+    let opts = Opts::new(name, help);
+    let counter = IntCounter::with_opts(opts)?;
+    prometheus::register(Box::new(counter.clone()))?;
+    Ok(counter)
+}
+
+/// Attempts to crate an `IntGauge`, returning `Err` if the registry does not accept the counter
+/// (potentially due to naming conflict).
+pub fn try_create_int_gauge(name: &str, help: &str) -> Result<IntGauge> {
+    let opts = Opts::new(name, help);
+    let gauge = IntGauge::with_opts(opts)?;
+    prometheus::register(Box::new(gauge.clone()))?;
+    Ok(gauge)
+}
+
+/// Attempts to crate a `Histogram`, returning `Err` if the registry does not accept the counter
+/// (potentially due to naming conflict).
+pub fn try_create_histogram(name: &str, help: &str) -> Result<Histogram> {
+    let opts = HistogramOpts::new(name, help);
+    let histogram = Histogram::with_opts(opts)?;
+    prometheus::register(Box::new(histogram.clone()))?;
+    Ok(histogram)
+}
+
+/// Starts a timer for the given `Histogram`, stopping when it gets dropped or given to `stop_timer(..)`.
+pub fn start_timer(histogram: &Result<Histogram>) -> Option<HistogramTimer> {
+    if let Ok(histogram) = histogram {
```

Comment:
This and other methods should do something meaningful when `Result` is `Err`. When is it going to fail? I suspect the failure means some issues with initializing the metrics reporting system, should we log the error and proceed? Either way, the error should be processed.